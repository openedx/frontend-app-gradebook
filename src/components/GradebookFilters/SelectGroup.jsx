/* eslint-disable react/sort-comp, react/button-has-type */
import React from 'react';
import PropTypes from 'prop-types';

import { Form } from '@edx/paragon';

const SelectGroup = ({
  id,
  label,
  value,
  onChange,
  disabled,
  options,
}) => (
  <div className="student-filters">
    <Form.Group controlId={id}>
      <Form.Label>{label}</Form.Label>
      <Form.Control as="select" {...{ value, onChange, disabled }}>
        {options}
      </Form.Control>
    </Form.Group>
  </div>
);
SelectGroup.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.node.isRequired,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  options: PropTypes.arrayOf(PropTypes.node).isRequired,
};
SelectGroup.defaultProps = {
  disabled: false,
};

export default SelectGroup;
