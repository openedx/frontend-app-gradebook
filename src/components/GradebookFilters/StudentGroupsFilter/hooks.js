import { actions, selectors, thunkActions } from 'data/redux/hooks';

export const useStudentGroupsFilterData = ({ updateQueryParams }) => {
  const selectedCohortEntry = selectors.root.useSelectedCohortEntry();
  const selectedTrackEntry = selectors.root.useSelectedTrackEntry();

  const cohorts = selectors.cohorts.useAllCohorts();
  const tracks = selectors.tracks.useAllTracks();

  const updateCohort = actions.filters.useUpdateCohort();
  const updateTrack = actions.filters.useUpdateTrack();

  const fetchGrades = thunkActions.grades.useFetchGrades();

  const handleUpdateTrack = (event) => {
    const selectedTrackItem = tracks.find(track => track.slug === event.target.value);
    const track = selectedTrackItem ? selectedTrackItem.slug.toString() : null;
    updateQueryParams({ track });
    updateTrack(track);
    fetchGrades();
  };

  const handleUpdateCohort = (event) => {
    const selectedCohortItem = cohorts.find(cohort => cohort.id === parseInt(event.target.value, 10));
    const cohort = selectedCohortItem ? selectedCohortItem.id.toString() : null;
    // the param expected to be cohort_id
    updateQueryParams({ cohort });
    updateCohort(cohort);
    fetchGrades();
  };
  return {
    cohorts: {
      value: selectedCohortEntry?.id || '',
      isDisabled: cohorts.length === 0,
      handleChange: handleUpdateCohort,
      entries: cohorts.map(({ id: value, name }) => ({ value, name })),
    },
    tracks: {
      value: selectedTrackEntry?.slug || '',
      handleChange: handleUpdateTrack,
      entries: tracks.map(({ slug: value, name }) => ({ value, name })),
    },
  };
};

export default useStudentGroupsFilterData;
