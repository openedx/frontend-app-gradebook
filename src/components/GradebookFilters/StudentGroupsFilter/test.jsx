/* eslint-disable import/no-named-as-default */

import React from 'react';
import { shallow } from 'enzyme';

import { fetchGrades } from 'data/thunkActions/grades';
import actions from 'data/actions';
import selectors from 'data/selectors';
import {
  optionFactory,
  StudentGroupsFilter,
  mapStateToProps,
  mapDispatchToProps,
} from '.';

jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    root: {
      selectedCohortEntry: jest.fn(state => ({ selectedCohortEntry: state })),
      selectedTrackEntry: jest.fn(state => ({ selectedTrackEntry: state })),
    },
    cohorts: {
      allCohorts: jest.fn(state => ({ allCohorts: state })),
      cohortsByName: jest.fn(state => ({ cohortsByName: state })),
    },
    tracks: {
      allTracks: jest.fn(state => ({ allTracks: state })),
      tracksByName: jest.fn(state => ({ tracksByName: state })),
    },
  },
}));

jest.mock('data/thunkActions/grades', () => ({
  fetchGrades: jest.fn(),
}));

describe('StudentGroupsFilter', () => {
  let props = {
    cohorts: [
      { name: 'cohorT1', id: 8001 },
      { name: 'cohorT2', id: 8002 },
      { name: 'cohorT3', id: 8003 },
    ],
    tracks: [
      { name: 'TracK1', slug: 'TracK1_slug' },
      { name: 'TracK2', slug: 'TracK2_slug' },
      { name: 'TRACK3', slug: 'TRACK3_slug' },
    ],
  };

  describe('optionFactory', () => {
    it('returns a list of options with a default first entry', () => {
      const data = [{ cMark: 'rainbow', name: 'RDash' }, { cMark: 'balloons', name: 'PPie' }];
      const defaultOption = 'All-Ponies';
      const key = 'cMark';
      const options = optionFactory({ data, defaultOption, key });
      expect(options).toMatchSnapshot();
    });
  });

  describe('Component', () => {
    beforeEach(() => {
      props = {
        ...props,
        intl: { formatMessage: (msg) => msg.defaultMessage },
        cohortsByName: {
          [props.cohorts[0].name]: props.cohorts[0],
          [props.cohorts[1].name]: props.cohorts[1],
          [props.cohorts[2].name]: props.cohorts[2],
        },
        tracksByName: {
          [props.tracks[0].name]: props.tracks[0],
          [props.tracks[1].name]: props.tracks[1],
          [props.tracks[2].name]: props.tracks[2],
        },
        fetchGrades: jest.fn(),
        selectedCohortEntry: props.cohorts[2],
        selectedTrackEntry: props.tracks[1],
        updateQueryParams: jest.fn(),
        updateCohort: jest.fn().mockName('updateCohort'),
        updateTrack: jest.fn().mockName('updateTrack'),
      };
    });

    describe('snapshots', () => {
      let el;
      beforeEach(() => {
        el = shallow(<StudentGroupsFilter {...props} />);
      });
      test('basic snapshot', () => {
        el.instance().updateTracks = jest.fn().mockName(
          'updateTracks',
        );
        el.instance().updateCohorts = jest.fn().mockName(
          'updateCohorts',
        );
        expect(el.instance().render()).toMatchSnapshot();
      });
      test('Cohorts group disabled if no cohorts', () => {
        el.setProps({ cohorts: [] });
        expect(el.instance().render()).toMatchSnapshot();
      });
      describe('mapCohortsEntries', () => {
        test('cohort options: [Cohort-All, <{slug, name}...>]', () => {
          expect(el.instance().mapCohortsEntries()).toMatchSnapshot();
        });
      });
      describe('mapTracksEntries', () => {
        test('cohort options: [Track-All, <{id, name}...>]', () => {
          expect(el.instance().mapTracksEntries()).toMatchSnapshot();
        });
      });
    });

    describe('behavior', () => {
      let el;
      beforeEach(() => {
        el = shallow(<StudentGroupsFilter {...props} />);
      });
      describe('selectedCohortIdFromEvent', () => {
        it('returns the id of the cohort with the name matching the event', () => {
          expect(
            el.instance().selectedCohortIdFromEvent(
              { target: { value: props.cohorts[1].name } },
            ),
          ).toEqual(props.cohorts[1].id.toString());
        });
        it('returns null if no matching cohort is found', () => {
          expect(
            el.instance().selectedCohortIdFromEvent(
              { target: { value: 'FAKE' } },
            ),
          ).toEqual(null);
        });
      });
      describe('selectedTrackSlugFromEvent', () => {
        it('returns the slug of the track with the name matching the event', () => {
          expect(
            el.instance().selectedTrackSlugFromEvent(
              { target: { value: props.tracks[1].name } },
            ),
          ).toEqual(props.tracks[1].slug);
        });
        it('returns null if no matching track is found', () => {
          expect(
            el.instance().selectedTrackSlugFromEvent(
              { target: { value: 'FAKE' } },
            ),
          ).toEqual(null);
        });
      });
      describe('updateTracks', () => {
        const selectedSlug = 'SLUG';
        beforeEach(() => {
          el = shallow(<StudentGroupsFilter {...props} />);
          jest.spyOn(
            el.instance(),
            'selectedTrackSlugFromEvent',
          ).mockReturnValue(selectedSlug);
          el.instance().updateTracks({ target: {} });
        });
        it('calls updateTrack with new value', () => {
          expect(props.updateTrack).toHaveBeenCalledWith(selectedSlug);
        });
        it('calls fetchGrades', () => {
          expect(props.fetchGrades).toHaveBeenCalledWith();
        });
        it('updates queryParams with track value', () => {
          expect(props.updateQueryParams).toHaveBeenCalledWith({
            track: selectedSlug,
          });
        });
      });
      describe('updateCohorts', () => {
        const selectedId = 23;
        beforeEach(() => {
          el = shallow(<StudentGroupsFilter {...props} />);
          jest.spyOn(
            el.instance(),
            'selectedCohortIdFromEvent',
          ).mockReturnValue(selectedId);
          el.instance().updateCohorts({ target: {} });
        });
        it('calls updateCohort with new value', () => {
          expect(props.updateCohort).toHaveBeenCalledWith(selectedId);
        });
        it('calls fetchGrades', () => {
          expect(props.fetchGrades).toHaveBeenCalledWith();
        });
        it('updates queryParams with cohort value', () => {
          expect(props.updateQueryParams).toHaveBeenCalledWith({
            cohort: selectedId,
          });
        });
      });
    });
  });
  describe('mapStateToProps', () => {
    const testState = { h: 'e', l: 'l', o: 'oooooooooo' };
    let mappedProps;
    beforeAll(() => {
      mappedProps = mapStateToProps(testState);
    });
    test('cohorts from selectors.cohorts.allCohorts', () => {
      expect(mappedProps.cohorts).toEqual(selectors.cohorts.allCohorts(testState));
    });
    test('cohortsByName from selectors.cohorts.cohortsByName', () => {
      expect(mappedProps.cohortsByName).toEqual(selectors.cohorts.cohortsByName(testState));
    });
    test('selectedCohortEntry from selectors.root.selectedCohortEntry', () => {
      expect(
        mappedProps.selectedCohortEntry,
      ).toEqual(selectors.root.selectedCohortEntry(testState));
    });
    test('selectedTrackEntry from selectors.root.selectedTrackEntry', () => {
      expect(
        mappedProps.selectedTrackEntry,
      ).toEqual(selectors.root.selectedTrackEntry(testState));
    });
    test('tracks from selectors.tracks.allTracks', () => {
      expect(mappedProps.tracks).toEqual(selectors.tracks.allTracks(testState));
    });
    test('tracksByName from selectors.tracks.tracksByName', () => {
      expect(mappedProps.tracksByName).toEqual(selectors.tracks.tracksByName(testState));
    });
  });
  describe('mapDispatchToProps', () => {
    test('fetchGrades from thunkActions.grades.fetchGrades', () => {
      expect(mapDispatchToProps.fetchGrades).toEqual(fetchGrades);
    });
    test('updateCohort from actions.filters.update.cohort', () => {
      expect(mapDispatchToProps.updateCohort).toEqual(actions.filters.update.cohort);
    });
    test('updateTrack from actions.filters.update.track', () => {
      expect(mapDispatchToProps.updateTrack).toEqual(actions.filters.update.track);
    });
  });
});
