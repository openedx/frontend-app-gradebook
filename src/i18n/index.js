import { messages as footerMessages } from '@edx/frontend-component-footer';
import { messages as headerMessages } from '@edx/frontend-component-header';
import { messages as paragonMessages } from '@edx/paragon';

import arMessages from './messages/ar.json';
import deMessages from './messages/de.json';
import es419Messages from './messages/es_419.json';
import faIRMessages from './messages/fa_IR.json';
import frMessages from './messages/fr.json';
import frCAMessages from './messages/fr_CA.json';
import hiMessages from './messages/hi.json';
import itMessages from './messages/it.json';
import ptMessages from './messages/pt.json';
import ruMessages from './messages/ru.json';
import ukMessages from './messages/uk.json';
import zhcnMessages from './messages/zh_CN.json';
// no need to import en messages-- they are in the defaultMessage field

const appMessages = {
  ar: arMessages,
  'es-419': es419Messages,
  'fa-ir': faIRMessages,
  fr: frMessages,
  'zh-cn': zhcnMessages,
  pt: ptMessages,
  it: itMessages,
  de: deMessages,
  hi: hiMessages,
  'fr-ca': frCAMessages,
  ru: ruMessages,
  uk: ukMessages,
};

export default [
  paragonMessages,
  footerMessages,
  headerMessages,
  appMessages,
];
