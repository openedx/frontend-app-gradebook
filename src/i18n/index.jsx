import arMessages from './messages/ar.json';
// no need to import en messages-- they are in the defaultMessage field
import es419Messages from './messages/es_419.json';
import frMessages from './messages/fr.json';
import zhcnMessages from './messages/zh_CN.json';

const messages = {
  ar: arMessages,
  'es-419': es419Messages,
  fr: frMessages,
  'zh-cn': zhcnMessages,
};

export default messages;
