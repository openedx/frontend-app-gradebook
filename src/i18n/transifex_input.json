{
  "gradebook.BulkManagementTab.csvUploadLabel": "Upload Grade CSV",
  "gradebook.BulkManagementTab.heading": "Use this feature by downloading a CSV for bulk management, overriding grades locally, and coming back here to upload.",
  "gradebook.BulkManagementTab.hint1": "Results appear in the table below.",
  "gradebook.BulkManagementTab.hint2": "Grade processing may take a few seconds.",
  "gradebook.BulkManagementTab.importBtnText": "Import Grades",
  "gradebook.BulkManagementTab.successDialog": "CSV processing. File uploads may take several minutes to complete."
}