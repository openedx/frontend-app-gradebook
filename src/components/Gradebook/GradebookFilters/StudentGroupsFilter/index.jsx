/* eslint-disable react/sort-comp, react/button-has-type, import/no-named-as-default */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { fetchGrades } from 'data/actions/grades';
import selectors from 'data/selectors';
import SelectGroup from '../SelectGroup';

export class StudentGroupsFilter extends React.Component {
  constructor(props) {
    super(props);
    this.updateCohorts = this.updateCohorts.bind(this);
    this.updateTracks = this.updateTracks.bind(this);
  }

  mapCohortsEntries = () => {
    const mapper = ({ id, name }) => (
      <option key={id} value={name}>{name}</option>
    );
    return [
      <option value="Cohort-All" key="0">Cohort-All</option>,
      ...this.props.cohorts.map(mapper),
    ];
  };

  mapTracksEntries = () => {
    const mapper = ({ slug, name }) => (
      <option key={slug} value={name}>{name}</option>
    );
    return [
      <option value="Track-All" key="0">Track-All</option>,
      ...this.props.tracks.map(mapper),
    ];
  };

  mapSelectedCohortEntry = () => {
    const selectedCohortEntry = this.props.cohorts.find(
      (x) => x.id === parseInt(this.props.selectedCohort, 10),
    );
    return selectedCohortEntry ? selectedCohortEntry.name : 'Cohorts';
  };

  mapSelectedTrackEntry = () => {
    const selectedTrackEntry = this.props.tracks.find(
      ({ slug }) => slug === this.props.selectedTrack,
    );
    return selectedTrackEntry ? selectedTrackEntry.name : 'Tracks';
  };

  selectedTrackSlugFromEvent(event) {
    const selectedTrackItem = this.props.tracks.find(
      ({ name }) => name === event.target.value,
    );
    return selectedTrackItem ? selectedTrackItem.slug : null;
  }

  selectedCohortIdFromEvent(event) {
    const selectedCohortItem = this.props.cohorts.find(
      x => x.name === event.target.value,
    );
    return selectedCohortItem ? selectedCohortItem.id.toString() : null;
  }

  updateTracks(event) {
    const selectedTrackSlug = this.selectedTrackSlugFromEvent(event);
    this.props.getUserGrades(
      this.props.courseId,
      this.props.selectedCohort,
      selectedTrackSlug,
      this.props.selectedAssignmentType,
    );
    this.props.updateQueryParams({ track: selectedTrackSlug });
  }

  updateCohorts(event) {
    const selectedCohortId = this.selectedCohortIdFromEvent(event);
    this.props.getUserGrades(
      this.props.courseId,
      selectedCohortId,
      this.props.selectedTrack,
      this.props.selectedAssignmentType,
    );
    this.props.updateQueryParams({ cohort: selectedCohortId });
  }

  render() {
    return (
      <>
        <SelectGroup
          id="Tracks"
          label="Tracks"
          value={this.mapSelectedTrackEntry()}
          onChange={this.updateTracks}
          options={this.mapTracksEntries()}
        />
        <SelectGroup
          id="Cohorts"
          label="Cohorts"
          value={this.mapSelectedCohortEntry()}
          disabled={this.props.cohorts.length === 0}
          onChange={this.updateCohorts}
          options={this.mapCohortsEntries()}
        />
      </>
    );
  }
}

StudentGroupsFilter.defaultProps = {
  cohorts: [],
  courseId: '',
  selectedAssignmentType: '',
  selectedCohort: null,
  selectedTrack: null,
  tracks: [],
};

StudentGroupsFilter.propTypes = {
  courseId: PropTypes.string,
  updateQueryParams: PropTypes.func.isRequired,

  // redux
  cohorts: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    id: PropTypes.number,
  })),
  getUserGrades: PropTypes.func.isRequired,
  selectedAssignmentType: PropTypes.string,
  selectedCohort: PropTypes.string,
  selectedTrack: PropTypes.string,
  tracks: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    slug: PropTypes.string,
  })),
};

export const mapStateToProps = (state) => {
  const { filters, cohorts, tracks } = selectors;
  return {
    cohorts: cohorts.allCohorts(state),
    selectedAssignmentType: filters.assignmentType(state),
    selectedCohort: filters.cohort(state),
    selectedTrack: filters.track(state),
    tracks: tracks.allTracks(state),
  };
};

export const mapDispatchToProps = {
  getUserGrades: fetchGrades,
};

export default connect(mapStateToProps, mapDispatchToProps)(StudentGroupsFilter);
