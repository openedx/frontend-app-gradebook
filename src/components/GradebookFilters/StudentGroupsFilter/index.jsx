/* eslint-disable react/sort-comp, react/button-has-type, import/no-named-as-default */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';

import actions from 'data/actions';
import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';

import messages from '../messages';
import SelectGroup from '../SelectGroup';

export const optionFactory = ({ data, defaultOption, key }) => [
  <option value={defaultOption} key="0">{defaultOption}</option>,
  ...data.map(
    entry => (<option key={entry[key]} value={entry.name}>{entry.name}</option>),
  ),
];

export class StudentGroupsFilter extends React.Component {
  constructor(props) {
    super(props);
    this.mapCohortsEntries = this.mapCohortsEntries.bind(this);
    this.mapTracksEntries = this.mapTracksEntries.bind(this);
    this.updateCohorts = this.updateCohorts.bind(this);
    this.updateTracks = this.updateTracks.bind(this);
  }

  mapCohortsEntries() {
    return optionFactory({
      data: this.props.cohorts,
      defaultOption: this.translate(messages.cohortAll),
      key: 'id',
    });
  }

  mapTracksEntries() {
    return optionFactory({
      data: this.props.tracks,
      defaultOption: this.translate(messages.trackAll),
      key: 'slug',
    });
  }

  selectedTrackSlugFromEvent({ target: { value } }) {
    const selectedTrackItem = this.props.tracksByName[value];
    return selectedTrackItem ? selectedTrackItem.slug : null;
  }

  selectedCohortIdFromEvent({ target: { value } }) {
    const selectedCohortItem = this.props.cohortsByName[value];
    return selectedCohortItem ? selectedCohortItem.id.toString() : null;
  }

  updateTracks(event) {
    const track = this.selectedTrackSlugFromEvent(event);
    this.props.updateQueryParams({ track });
    this.props.updateTrack(track);
    this.props.fetchGrades();
  }

  updateCohorts(event) {
    const cohort = this.selectedCohortIdFromEvent(event);
    this.props.updateQueryParams({ cohort });
    this.props.updateCohort(cohort);
    this.props.fetchGrades();
  }

  translate(message) {
    return this.props.intl.formatMessage(message);
  }

  render() {
    return (
      <>
        <SelectGroup
          id="Tracks"
          label={this.translate(messages.tracks)}
          value={this.props.selectedTrackEntry.name}
          onChange={this.updateTracks}
          options={this.mapTracksEntries()}
        />
        <SelectGroup
          id="Cohorts"
          label={this.translate(messages.cohorts)}
          value={this.props.selectedCohortEntry.name}
          disabled={this.props.cohorts.length === 0}
          onChange={this.updateCohorts}
          options={this.mapCohortsEntries()}
        />
      </>
    );
  }
}

StudentGroupsFilter.defaultProps = {
  cohorts: [],
  cohortsByName: {},
  selectedCohortEntry: { name: '' },
  selectedTrackEntry: { name: '' },
  tracks: [],
  tracksByName: {},
};

StudentGroupsFilter.propTypes = {
  updateQueryParams: PropTypes.func.isRequired,

  // injected
  intl: intlShape.isRequired,

  // redux
  cohorts: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    id: PropTypes.number,
  })),
  cohortsByName: PropTypes.objectOf(PropTypes.shape({
    name: PropTypes.string,
    id: PropTypes.number,
  })),
  fetchGrades: PropTypes.func.isRequired,
  selectedTrackEntry: PropTypes.shape({ name: PropTypes.string }),
  selectedCohortEntry: PropTypes.shape({ name: PropTypes.string }),
  tracks: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    slug: PropTypes.string,
  })),
  tracksByName: PropTypes.objectOf(PropTypes.shape({
    name: PropTypes.string,
    slug: PropTypes.string,
  })),
  updateCohort: PropTypes.func.isRequired,
  updateTrack: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  cohorts: selectors.cohorts.allCohorts(state),
  cohortsByName: selectors.cohorts.cohortsByName(state),
  selectedCohortEntry: selectors.root.selectedCohortEntry(state),
  selectedTrackEntry: selectors.root.selectedTrackEntry(state),
  tracks: selectors.tracks.allTracks(state),
  tracksByName: selectors.tracks.tracksByName(state),
});

export const mapDispatchToProps = {
  fetchGrades: thunkActions.grades.fetchGrades,
  updateCohort: actions.filters.update.cohort,
  updateTrack: actions.filters.update.track,
};

export default injectIntl(connect(mapStateToProps, mapDispatchToProps)(StudentGroupsFilter));
