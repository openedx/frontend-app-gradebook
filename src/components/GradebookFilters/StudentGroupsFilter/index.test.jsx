import React from 'react';
import { shallow } from 'enzyme';
import { useIntl } from '@edx/frontend-platform/i18n';

import SelectGroup from '../SelectGroup';
import useStudentGroupsFilterData from './hooks';
import StudentGroupsFilter from '.';

jest.mock('../SelectGroup', () => 'SelectGroup');
jest.mock('./hooks', () => ({ __esModule: true, default: jest.fn() }));

const props = {
  cohorts: {
    value: 'test-cohort',
    entries: [
      { value: 'v1', name: 'n1' },
      { value: 'v2', name: 'n2' },
      { value: 'v3', name: 'n3' },
    ],
    handleChange: jest.fn(),
    isDisabled: false,
  },
  tracks: {
    value: 'test-track',
    entries: [
      { value: 'v1', name: 'n1' },
      { value: 'v2', name: 'n2' },
      { value: 'v3', name: 'n3' },
      { value: 'v4', name: 'n4' },
    ],
    handleChange: jest.fn(),
  },
};
useStudentGroupsFilterData.mockReturnValue(props);
const updateQueryParams = jest.fn();

let el;
describe('StudentGroupsFilter component', () => {
  beforeAll(() => {
    jest.clearAllMocks();
    el = shallow(<StudentGroupsFilter updateQueryParams={updateQueryParams} />);
  });
  describe('behavior', () => {
    it('initializes hooks', () => {
      expect(useStudentGroupsFilterData).toHaveBeenCalledWith({ updateQueryParams });
      expect(useIntl).toHaveBeenCalledWith();
    });
  });
  describe('render', () => {
    test('snapshot', () => {
      expect(el).toMatchSnapshot();
    });
    test('track options', () => {
      const {
        options,
        onChange,
        value,
      } = el.find(SelectGroup).at(0).props();
      expect(value).toEqual(props.tracks.value);
      expect(onChange).toEqual(props.tracks.handleChange);
      expect(options.length).toEqual(5);
      const testEntry = props.tracks.entries[0];
      const optionProps = options[1].props;
      expect(optionProps.value).toEqual(testEntry.value);
      expect(optionProps.children).toEqual(testEntry.name);
    });
    test('cohort options', () => {
      const {
        options,
        onChange,
        disabled,
        value,
      } = el.find(SelectGroup).at(1).props();
      expect(value).toEqual(props.cohorts.value);
      expect(disabled).toEqual(false);
      expect(onChange).toEqual(props.cohorts.handleChange);
      expect(options.length).toEqual(4);
      const testEntry = props.cohorts.entries[0];
      const optionProps = options[1].props;
      expect(optionProps.value).toEqual(testEntry.value);
      expect(optionProps.children).toEqual(testEntry.name);
    });
  });
});
