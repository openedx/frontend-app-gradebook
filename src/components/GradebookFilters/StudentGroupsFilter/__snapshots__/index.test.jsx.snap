// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentGroupsFilter component render snapshot 1`] = `
<Fragment>
  <SelectGroup
    id="Tracks"
    label="Tracks"
    onChange={[MockFunction]}
    options={
      Array [
        <option
          value="Track-All"
        >
          Track-All
        </option>,
        <option
          value="v1"
        >
          n1
        </option>,
        <option
          value="v2"
        >
          n2
        </option>,
        <option
          value="v3"
        >
          n3
        </option>,
        <option
          value="v4"
        >
          n4
        </option>,
      ]
    }
    value="test-track"
  />
  <SelectGroup
    disabled={false}
    id="Cohorts"
    label="Cohorts"
    onChange={[MockFunction]}
    options={
      Array [
        <option
          value="Cohort-All"
        >
          Cohort-All
        </option>,
        <option
          value="v1"
        >
          n1
        </option>,
        <option
          value="v2"
        >
          n2
        </option>,
        <option
          value="v3"
        >
          n3
        </option>,
      ]
    }
    value="test-cohort"
  />
</Fragment>
`;
